### ISSUE TITLE
About the baselines

### ISSUE THREAD

**Issue by dongqian0206** [zdhNarsil/Diffusion-Generative-Flow-Samplers#1]
Hi Dinghuai,

Thank you for releasing the codes.

I have been working on reproducing the results based on your source codes. However, when running PIS and DDS (baseline), I obtained an average LogZ Error of around 0.4 and 0.3 for MoG, whereas the reported one in your paper is just around 0.036 and 0.028. I have attempted to run the official codes, but encountered a series of package issues that I am finding challenging to resolve.

Would you like to provide the baseline codes used for your paper as well? It would be helpful in resolving the discrepancies.

Best,

Dong

**Comment by zdhNarsil** [comment]
Hi, thanks for your interest, I wonder in PIS, have you set the dt to be 0.05? For DDS I think one also needs to change the sigma appropriately as a hyperparameter.

**Comment by dongqian0206** [comment]
Hi, 

Yes, I have set dt = 0.05 for MoG. After double-checking both your codes (as provided below) and mine for PIS. I noticed a difference - DGFS is updated sample-by-sample, each of (N+1, d), while in my implementation, PIS is updated based on all samples (bz, N+1, d). 

I am curious about the implementation on your end. Could you share some insights or considerations regarding this difference?

Thank you & regards.

 ```
for b_idx in range(batch_size):
      state_b = states[b_idx]  # (N+1, d)
      log_pf = self.log_pf(time_b[:-1], state_b[:-1], state_b[1:])  # (N,)
      log_pb = self.log_pb(time_b[1:], state_b[1:], state_b[:-1])  # (N,)

      flow_b = self.flow(time_b, state_b).squeeze(-1)  # (N+1, 1) -> (N+1,)

      flow_b = flow_b + logrs[b_idx]  # (N+1,)
      flow_b[-1] = logrs[b_idx][-1]  # (1,)

      diff_logp = log_pf - log_pb # (N, )
      diff_logp_padded = torch.cat((torch.zeros(1).to(diff_logp), diff_logp.cumsum(dim=-1)), dim=0) # (N+1,)
      # this means A1[i, j] = diff_logp[i:j].sum(dim=-1)
      A1 = diff_logp_padded.unsqueeze(0) - diff_logp_padded.unsqueeze(1)  # (N+1, N+1)

      A2 = flow_b[:, None] - flow_b[None, :] + A1  # (N+1, N+1)
      if not torch.all(torch.isfinite(A2)):
          import ipdb;
          ipdb.set_trace()

      A2 = (A2 / self.data_ndim).pow(2)  # (N+1, N+1)
      # torch.triu() is useless here
      loss = torch.triu(A2 * self.coef, diagonal=1).sum()
      info["loss_dlogp"] = loss.item()

      self.optimizer.zero_grad()
      loss.backward()
      self.optimizer.step()
```

**Comment by dongqian0206** [comment]
Hi,

Yes, I have set dt = 0.05 for MoG. After double-checking both your codes for DGFS (as provided below) and mine for PIS, I noticed two differences - 

1. DGFS is updated sample-by-sample, each with the size of (N + 1, dim), while in my implementation, PIS is updated by all mini-batches of samples with the size of (bz, N, dim). 
2. You used PF and PB, while I used (u2 + prior_nll - terminal_nll) for training PIS.

**Update:** I tried $min_{\theta} \log Q_{\theta} \left[ \log \frac{Q_{\theta}}{P} \right]$, where stop_gradient for $\log \frac{Q_{\theta}}{P}$ and followed your ways to define $Q_{\theta}$ and $P$ (PF and PB, respectively in your implementation). Now logZ = 0.0112 for one run. However, if I use $\log \frac{Q_{\theta}}{P} =$ (u2 + prior_nll - terminal_nll) as the objective, the performance was quite poor. Did you encounter such issue?

Thank you & Regards

**Comment by zdhNarsil** [comment]
Hi,

The two points do not matter much according to my experience. Do you write your own implementation of the baselines? I suspect that there could be small bugs in your implementation. Feel free to email me for more discussions on the baseline implementation -- it is actually not quite related to the implementation of this repo.
